View FileUpload / Modeling / Text submission from Exercise Scores page

madwau opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~Server: I added multiple integration tests (Spring) related to the features~
- [x] ~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~
- [x] ~Server: I implemented the changes with a good performance and prevented too many database calls~
- [x] ~Server: I documented the Java code using JavaDoc style.~
- [x] ~Client: I added multiple integration tests (Jest) related to the features~
- [x] ~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This PR addresses issue #1120 which states that it is not possible to reach the submission itself for TextExercise and FileUploadExercise exercises from the scores or participations page (i.e. not reachable via any link or button, only by manually extracting the submission id and inserting it into a url pattern in the browser url bar). For programming exercises this is possible by clicking on the "Repository" button or link.

### Description
Note: You need to be an instructor for that exercise.
1. Add button "View submission" to scores page for FileUploadExercise (opens the assessment page for the latest file upload submission)
2. Add button "View submission" to scores page for TextExercise (opens the assessment page for the latest text submission)
3. Add button "View submission" to scores page for ModelingExercise (opens the assessment page for the latest modeling submission)

### Steps for Testing
1. Log in to Artemis as an instructor
2. Navigate to Course Administration
3. Click on "Exercises" for a course with a file upload, modeling and text exercise
4. Click on "Scores" for a file upload, modeling and text exercise
5. Check that the new button (either "Download file" or "View Text Submission" or "View Modeling Submission") is present and performs its function as described above


### Screenshots
Note: Button text has changed to "View submission".

![OnPaste 20191215-143754](https://user-images.githubusercontent.com/7109225/70863452-93713e00-1f48-11ea-8b61-50853ca178bb.png)
![OnPaste 20191215-144113](https://user-images.githubusercontent.com/7109225/70863481-f662d500-1f48-11ea-9a44-ae7948ca6af5.png)
![OnPaste 20191215-153608](https://user-images.githubusercontent.com/7109225/70864245-9ff99480-1f50-11ea-9e42-1d1d6535ec7b.png)

madwau commented:
> Thanks, I don't have any real issues. We might want to also open the assessment page for the file upload exercise in addition to downloading the file, just to be consistent with the other types. What do you think?

I thought about that as well and wasn't sure. It would be good for consistency but would increase the number of clicks needed to see the actual submission content by one. Let's maybe see what the other reviewers think.

krusche commented on src/main/webapp/app/scores/exercise-scores.component.ts:
```diff
@@ -34,9 +36,12 @@ export class ExerciseScoresComponent implements OnInit, OnDestroy {
-    readonly PROGRAMMING = ExerciseType.PROGRAMMING;
+    readonly FILE_UPLOAD = ExerciseType.FILE_UPLOAD;
     readonly MODELING = ExerciseType.MODELING;
+    readonly PROGRAMMING = ExerciseType.PROGRAMMING;
```
you can also just use 
`readonly ExerciseType = ExerciseType;`

then in the html, you use ExerciseType.PROGRAMMING or ExerciseType.MODELING

madwau replied:
True, that's also how it is done in the participation component. I refactored it.

krusche commented on src/main/webapp/i18n/de/modelingSubmission.json:
```diff
@@ -15,6 +15,7 @@
             "detail": {
                 "title": "Modeling Submission"
             },
+            "view": "Modeling Submission ansehen",
```
please add proper translation here. In German we use:
Modeling --> Modellierung
Submission --> Einreichung

krusche replied:
and please also change all other wrong values in this file

madwau replied:
I updated the translations as discussed for all the submission types. Instead of the full variant "Modellierungseinreichung" or "Texteinreichung", the German language version now simply uses "Einreichung" since it is simpler and its meaning should be clear from context anyway.

krusche commented on src/main/webapp/i18n/de/textSubmission.json:
```diff
@@ -15,7 +15,8 @@
             },
-            "text": "Text"
+            "text": "Text",
+            "view": "Text Submission ansehen"
```
Submission --> Einreichung

(please also change all other wrong values in this file)

madwau replied:
See thread above.

clemenszuck commented:
During testing I noticed that for the case of Text exercises, after you click on "View Text Submission" the Back button is visible on the Assessment page: 

![image](https://user-images.githubusercontent.com/33415047/70915819-dac7ff00-201a-11ea-8e87-c5586927cbef.png)

If you click on it you are directed to the Tutor Dashboard of that exercise. I think thats pretty confusing and the Back button should be not visible, as it is on the Modeling Assessment page. Or is there any reason for it to be visible on the Text assessment, but not on the Modeling assessment page?

Everything else works as expected!

ungaralex commented:
> During testing I noticed that for the case of Text exercises, after you click on "View Text Submission" the Back button is visible on the Assessment page:
> 
> ![image](https://user-images.githubusercontent.com/33415047/70915819-dac7ff00-201a-11ea-8e87-c5586927cbef.png)
> 
> If you click on it you are directed to the Tutor Dashboard of that exercise. I think thats pretty confusing and the Back button should be not visible, as it is on the Modeling Assessment page. Or is there any reason for it to be visible on the Text assessment, but not on the Modeling assessment page?
> 
> Everything else works as expected!

We have the same behavior for other components. It's not pretty, but also nothing that got introduced with this PR. We already talked about this once (couple of weeks ago) if I remember correctly and were discussing whether we should remove this back button altogether.

clemenszuck commented:
> We have the same behavior for other components. It's not pretty, but also nothing that got introduced with this PR. We already talked about this once (couple of weeks ago) if I remember correctly and were discussing whether we should remove this back button altogether.

Okay, makes sense, not in the scope of this PR then.

madwau commented:
> During testing I noticed that for the case of Text exercises, after you click on "View Text Submission" the Back button is visible on the Assessment page:
> 
> ![image](https://user-images.githubusercontent.com/33415047/70915819-dac7ff00-201a-11ea-8e87-c5586927cbef.png)
> 
> If you click on it you are directed to the Tutor Dashboard of that exercise. I think thats pretty confusing and the Back button should be not visible, as it is on the Modeling Assessment page. Or is there any reason for it to be visible on the Text assessment, but not on the Modeling assessment page?
> 
> Everything else works as expected!

Thanks for the feedback! Yes, the back button does unfortunately behave strangely in this case. For modeling exercise submissions a query param `hideBackButton` can be specified to achieve the desired behavior but this is the only exercise type where this is currently supported as far as I know.

As @ungaralex said, this can be tackled in another PR (either by removing the back button all together or making it optional when using the respective components).

madwau commented on src/main/webapp/app/scores/exercise-scores.component.html:
```diff
@@ -173,21 +177,24 @@ <h2>
                     <ng-template ngx-datatable-cell-template let-value="value">
-                        <button *ngIf="exercise.type === PROGRAMMING" class="btn btn-primary btn-sm mr-1" (click)="goToRepository(value)">
-                            <fa-icon [icon]="'code-branch'" class="mr-1"></fa-icon>Repository
+                        <button *ngIf="exercise.type === ExerciseType.PROGRAMMING" class="btn btn-primary btn-sm mr-1" (click)="goToRepository(value)">
```
I think the user's main wish was that he can quickly get from the result to the submission content itself (file, diagram, text, etc). But since this leads to the assessment page for most exercise types, we should probably also link to the assessment page for a file upload exercise (and as you said, there will probably be users who appreciate the possibility to access the assessment page and not just the file itself). It will be one more click to see the file content itself but that should be tolerable.

madwau replied:
Implemented the change for the file upload exercise.

madwau commented:
> Looks good to me! Maybe just update your PR description as it says now "view submission" for all exercise types

Thanks, I updated the description and added a note above the screenshots.

krusche commented on src/main/webapp/app/scores/exercise-scores.component.html:
```diff
@@ -198,6 +205,37 @@ <h2>
+
+                <ngx-datatable-column prop="" *ngIf="exercise.type === ExerciseType.TEXT" [resizeable]="false" [minWidth]="180" [width]="180">
+                    <ng-template ngx-datatable-cell-template let-value="value">
+                        <button routerLink="/text/{{ exercise.id }}/assessment/{{ value.submission.id }}" class="btn btn-primary btn-sm mr-1">
```
Why are you navigating to the assessment page here, but for the other 2 exercise types, you navigate to the submission page?

I think this should be consistent for all exercise types! 

We also need to make sure that tutors can not navigate to arbitrary assessments on this page because this might be problematic for our double blind assessment mechanism

madwau replied:
As discussed, all of them lead to the assessment page of the submission for now (in the future they might lead to a read only version of only the submission) and the buttons will only be shown to instructors of that exercise.

madwau replied:
I just tested the instructor only display of the new "View submission" buttons on the test server but maybe someone else can verify it as well.

madwau commented:
> Nice work! Except for the weird back button behavior, I did not find problems.
> 
> I would suggest that we refactor and do not remove the back button functionality, but this is a discussion for another issue.

Thanks! Yes, we should discuss that in the corresponding issue.
